### ISSUE TITLE
Have a way to go back to the last valid state

### ISSUE THREAD

**Issue by deanljohnson** [Seanv554/leapmotion_boardgame#3]
Users are undoubtedly going to bump pieces too hard or accidentally send pieces flying, so whenever a piece is dropped in a valid location, we should save the state of the board so the players can always go back to that state.

**Comment by deanljohnson** [comment]
Even without having collisions between the hand and the pieces, the interaction can still sometimes behave unexpected, so this is still an issue. Could also be helpful as a type of fall back mechanism - if something really unexpected happens we can just call a function to revert the board to the previous state.

**Comment by seanvail** [comment]
Dean, gonna start working on this, just want to make sure you haven't started already.

**Comment by deanljohnson** [comment]
Nope, haven't done anything with this, though I was thinking the simplest way to handle this would be for all pieces to remember their previous position. Then when we want to go back to the previous state, we just check the pieces for their last position and place them there. There will probably have to be special care taken with pieces that are captured (not currently implemented) or when multiple-moves are made in one turn - such as in checkers where you can jump more than once in a turn.

**Comment by seanvail** [comment]
Save State function created. The function saves entries, each entry contains copies of the X, Y, isKing, & isBlack variables of a checkers piece, into a state object which is saved as a variable in CheckersBoard and will be used to reload the state in the future. Will add the load function tomorrow.

**Comment by deanljohnson** [comment]
Nice work. 

Just so you are aware, I recently committed some code that deletes pieces when they get jumped, but I made sure that the piece capturing happens after the save state call for that move, so your data will know where the captured piece was.

Also, I implemented turns, so you're going to need to save whose turn it is.

**Comment by seanvail** [comment]
Okay, so the load function is working...but, it looks like either the game rules and/or move validation (need your input to know for sure), are causing the pieces to all zoom into one area after I reload them. 

First screenshot: Taken shortly (few frames after) the pieces are deleted and then reloaded into their old positions above the board (Note: I'm using the exact same spawning logic as the SpawnPieces function).

![pull issue](https://cloud.githubusercontent.com/assets/5431312/14451005/1cc0b566-0038-11e6-87d5-8bf5008b74ef.png)

Seconds screenshot: Pieces all clustered around one corner. 

![pull issue 2](https://cloud.githubusercontent.com/assets/5431312/14451008/224d5e44-0038-11e6-8320-8f2bb5dee9ad.png)

What do you think is causing this?

**Comment by deanljohnson** [comment]
Ahh, that was something I just fixed when I was working on resetting the board after the game is won. It was due to pieces defaulting to index {0, 0} and being placed there. I already have committed the fix, so if you pull down the latest version you should be fine - merging is going to be kind of a pain because UE4 saves everything in binary format. The relevant changes I made were in the CheckersPiece and CheckersBoard blueprints - basically I set the x/y index of a piece when spawning it, rather then spawning and then setting the index.

**Comment by seanvail** [comment]
Yea I'm going to have to manually merge CheckersBoard, downside to
blueprints I suppose.

**Comment by deanljohnson** [comment]
Yep. I'll list the changes in CheckersBoard I made so it will hopefully be easier for you to do so. I'm assuming that you already merged CheckersPiece (I'm guessing there are no conflicts):

In SpawnPieces: with the two calls to SpawnActor, pass in the indices from the for loops into the x/y inputs. The top for loop is the y-index, bottom is x. Also, delete the Set X and Set Y nodes on the right side of the Spawn Red/Black commented boxes. And I just realized I committed a testing setup - in the box commented "Don't spawn in center area" (bottom left), change the values to 3 (top) and 4 (bottom). My apologies for that one.

CheckForGameOver: new function
EndGame: new function
MakeMove: there is a call to CheckForGameOver after the CapturePiece call, and that connects to a branch. On true, it calls EndGame and then returns false for moreJumps. On false, it does the same as before.

EventGraph: this is the sucky one. If GameOver (new variable) is true, I delete all the pieces, reset all variables, and then spawn new pieces.

**Comment by seanvail** [comment]
Okay, getting there, but there still a strange quirk:

When I reload the saved checkpoint (in this case I'm loading the default layout), a few specific pieces teleport from the place they're supposed to be in to this same corner again. Note: It's the same pieces every time.

![corner issue](https://cloud.githubusercontent.com/assets/5431312/14452733/789ea39a-0045-11e6-8189-4043659df283.png)

**Comment by deanljohnson** [comment]
Ahh figured it out - in your Save State function, you were adding an item to the array before getting it's index/isKing values, so I guess unreal was using default values for indices when it added them to the array. I really didn't think that would be the issue when I looked at it because it seems like it should affect all the pieces, but something told me to try it.

Notice that it only effected the last black/red piece to be spawned.... very strange off by one error I guess.

Anyways, I reordered it and now it works fine. Going to run a quick check through some of the blueprints and then I'll be pushing the bug-fix.
